Small interaction regions need to be guarded with occlusion regions.

megangardner opened the pull request:
#### e69ef3dfd95259622b14db520960c37959feeafd
<pre>
Small interaction regions need to be guarded with occlusion regions.
<a href="https://bugs.webkit.org/show_bug.cgi?id=254785">https://bugs.webkit.org/show_bug.cgi?id=254785</a>
rdar://105092716

Reviewed by Tim Horton.

Also add some additional debug information to draw occlusion regions
in red.

* LayoutTests/interaction-region/input-type-file-region-expected.txt:
* LayoutTests/interaction-region/input-type-range-region-expected.txt:
* Source/WebCore/page/DebugPageOverlays.cpp:
(WebCore::InteractionRegionOverlay::drawRect):
* Source/WebCore/rendering/EventRegion.cpp:
(WebCore::EventRegionContext::guardSmallIntetactionRegions):
(WebCore::EventRegionContext::copyInteractionRegionsToEventRegion):
* Source/WebCore/rendering/EventRegion.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::paintDebugOverlays):

Canonical link: <a href="https://commits.webkit.org/262655@main">https://commits.webkit.org/262655@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b5aea908663f10df05fc68d479c38daf21ebf9b8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/2163 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/2196 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/2264 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/3086 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/2205 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/2136 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/2276 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/2241 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/1958 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/2182 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/1931 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/1936 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/2941 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/1916 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/1916 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/1814 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/1994 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/1964 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/3102 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/1982 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/1776 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/1907 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/1926 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/535 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/2099 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebCore/rendering/EventRegion.cpp:
```diff
@@ -208,8 +208,44 @@ bool EventRegionContext::shouldConsolidateInteractionRegion(IntRect bounds, Rend
+void EventRegionContext::guardSmallIntetactionRegions()
+{
+    static int minimumSize = 20;
+    static int occulsionBuffer = 10;
```
s/buffer/margin/?

hortont424 replied:
s/occulsion/occlusion/ eeeeeeverywhere

hortont424 commented on Source/WebCore/rendering/EventRegion.h:
```diff
@@ -76,6 +76,7 @@ class EventRegionContext {
     HashSet<IntRect> m_interactionRects;
     HashSet<IntRect> m_occlusionRects;
     HashMap<ElementIdentifier, Region> m_discoveredRegionsByElement;
+    void guardSmallIntetactionRegions();
```
1) This is in the wrong place
2) Typo in intetaction
3) Maybe a slightly more wordy name

hortont424 commented on Source/WebCore/rendering/RenderLayerBacking.cpp:
```diff
@@ -3688,10 +3688,17 @@ void RenderLayerBacking::paintDebugOverlays(const GraphicsLayer* graphicsLayer,
 
 #if ENABLE(INTERACTION_REGIONS_IN_EVENT_REGION)
     if (DebugPageOverlays::shouldPaintOverlayIntoLayerForRegionType(renderer().page(), DebugPageOverlays::RegionType::InteractionRegion)) {
+        
```
but why?

pxlcoder commented on Source/WebCore/rendering/RenderLayerBacking.cpp:
```diff
@@ -3692,6 +3692,12 @@ void RenderLayerBacking::paintDebugOverlays(const GraphicsLayer* graphicsLayer,
         context.setStrokeThickness(1);
 
         for (const auto& region : eventRegion.interactionRegions()) {
+            
```
Unnecessary newline.

pxlcoder commented on Source/WebCore/rendering/RenderLayerBacking.cpp:
```diff
@@ -3692,6 +3692,12 @@ void RenderLayerBacking::paintDebugOverlays(const GraphicsLayer* graphicsLayer,
+            if (region.type == InteractionRegion::Type::Occlusion)
+                context.setStrokeColor(Color::red);
+            else
+                context.setStrokeColor(Color::green);
```
Since `Color::green` is always being set in `else`, you can remove the `context.setStrokeColor(Color::green);` on line 3691 above.

pxlcoder commented on Source/WebCore/rendering/EventRegion.cpp:
```diff
@@ -208,8 +208,44 @@ bool EventRegionContext::shouldConsolidateInteractionRegion(IntRect bounds, Rend
+void EventRegionContext::addOcclusionMarginsForSmallInteractionRegions()
+{
+    static int minimumSize = 20;
+    static int occlusionMargin = 10;
```
```suggestion
    constexpr int minimumSize = 20;
    constexpr int occlusionMargin = 10;
```

pxlcoder commented on Source/WebCore/rendering/EventRegion.cpp:
```diff
@@ -208,8 +208,44 @@ bool EventRegionContext::shouldConsolidateInteractionRegion(IntRect bounds, Rend
+    static int minimumSize = 20;
+    static int occlusionMargin = 10;
+    Vector<InteractionRegion> interactionRegionsWithOcclusions;
+    bool addedOcclutions = false;
```
This variable currently seems unnecessary, since you could just check `!interactionRegionsWithOcclusions.isEmpty()` in its place.

(it's also a misspelling as-is)

pxlcoder commented on Source/WebCore/rendering/EventRegion.cpp:
```diff
@@ -208,8 +208,44 @@ bool EventRegionContext::shouldConsolidateInteractionRegion(IntRect bounds, Rend
+                InteractionRegion::Type::Occlusion
+            });
+            addedOcclutions = true;
+            interactionRegionsWithOcclusions.append(interactionRegion);
```
I don't have full context into this change, but is it intentional that the original `interactionRegion` is only added if `isSmallRect` is true?

In the case where `m_interactionRegions` contains two regions, and one is deemed "small" and the other is not, the latter region would get dropped from `m_interactionRegions`. Is that scenario possible and is the described behavior desirable?

etiennesegonzac commented on Source/WebCore/rendering/EventRegion.cpp:
```diff
@@ -209,8 +209,42 @@ bool EventRegionContext::shouldConsolidateInteractionRegion(IntRect bounds, Rend
+    constexpr int minimumSize = 20;
+    constexpr int occlusionMargin = 10;
+    Vector<InteractionRegion> interactionRegionsWithOcclusions;
+    for (auto& interactionRegion : m_interactionRegions) {
```
we should only look at `Type::Interaction` here, no need to shield the existing occlusions.
(I get that we still need to append Occlusions to the final vector but I have another comment below :))

pxlcoder commented on Source/WebCore/rendering/EventRegion.cpp:
```diff
@@ -209,8 +209,42 @@ bool EventRegionContext::shouldConsolidateInteractionRegion(IntRect bounds, Rend
+        interactionRegionsWithOcclusions.append(interactionRegion);
+    }
+    if (!interactionRegionsWithOcclusions.isEmpty())
+        m_interactionRegions = interactionRegionsWithOcclusions;
```
I see you made the change I described here: https://github.com/WebKit/WebKit/pull/12220#discussion_r1153953919/.

Now, `!interactionRegionsWithOcclusions.isEmpty()` is always going to be true, when `m_interactionRegions` is non-empty, due to line 239. Can we just append directly to `m_interactionRegions`, rather than creating the temporary vector and copying, or does the order of interaction/occlusion region pairs matter?

etiennesegonzac commented on Source/WebCore/rendering/EventRegion.cpp:
```diff
@@ -209,8 +209,42 @@ bool EventRegionContext::shouldConsolidateInteractionRegion(IntRect bounds, Rend
+            interactionRegionsWithOcclusions.append({
+                interactionRegion.elementIdentifier,
+                occlusionRect,
+                interactionRegion.borderRadius,
```
not sure, but it might be counter productive to relay the border radius as it could reduce the positive impact of the "shield"

etiennesegonzac commented on Source/WebCore/rendering/EventRegion.cpp:
```diff
@@ -209,8 +209,42 @@ bool EventRegionContext::shouldConsolidateInteractionRegion(IntRect bounds, Rend
+        if (interactionRegion.rectInLayerCoordinates.height() < minimumSize) {
+            occlusionRect.inflateY(occlusionMargin);
+            isSmallRect = true;
+        }
```
since the conditions here are based on static values and the rect coordinates, wouldn't it be easier to do in as-we-go? [here](https://github.com/WebKit/WebKit/blob/a8fc73077082ccee3003ba4eaab65cfb23d9c37b/Source/WebCore/rendering/EventRegion.cpp#L151)

* we're already in an `Type::Interaction` branch
* we can append the shield first to get the expected order
* it'll save us one copy of the whole Vector

For the case where we have multiple rects we just need to make sure to append all shields (Occlusions) before we append the Interactions.

etiennesegonzac commented on Source/WebCore/rendering/EventRegion.cpp:
```diff
@@ -164,7 +197,14 @@ void EventRegionContext::uniteInteractionRegions(const Region& region, RenderObj
             m_occlusionRects.add(bounds);
         }
-
+        auto occlusionRect = guardRectForRegionBounds(interactionRegion->rectInLayerCoordinates);
```
we could have a `Type::Occlusion` here and don't want to generate a guard in that case.

the method is getting a bit complex anyway, plus we have a branche that always return but still gets an else...
I'd suggest starting with a `interactionRegion->type == InteractionRegion::Type::Occlusion` condition, where we handle the simpler occlusion case and early return.
and then we can have all the consolidation/discovered region/guard code (everything that applies to `Type::Interaction`).

etiennesegonzac commented on Source/WebCore/rendering/EventRegion.cpp:
```diff
@@ -119,6 +119,30 @@ bool EventRegionContext::contains(const IntRect& rect) const
 
 #if ENABLE(INTERACTION_REGIONS_IN_EVENT_REGION)
 
+static std::optional<IntRect> guardRectForRegionBounds(const IntRect& regionBounds)
```
thanks! I think it helps a lot to have this logic extracted here
nit: I'd group it with `shouldConsolidateInteractionRegion` and remove the last empty line

megangardner replied:
I didn't move it down since it's a static method and needs to be implemented or declared before it's used.

etiennesegonzac commented on LayoutTests/interaction-region/input-type-range-region-expected.txt:
```diff
@@ -12,6 +12,8 @@
 
       (interaction regions [
+        (occlusion (48,-8) width=37 height=36)
+        (borderRadius 0.00),
```
we are covering everything with these existing cases, but I'm afraid we could loose this coverage during an unrelated html forms change.

can we add a small guard-focused test?

it would also be a good place to test that we're **not** generating guards for small occlusion layers

etiennesegonzac commented on Source/WebCore/rendering/EventRegion.h:
```diff
@@ -67,6 +67,10 @@ class EventRegionContext {
 private:
+#if ENABLE(INTERACTION_REGIONS_IN_EVENT_REGION)
+    void addOcclusionMarginsForSmallInteractionRegions();
+#endif
```
not needed anymore

hortont424 commented on LayoutTests/interaction-region/tiny-regions-expected.txt:
```diff
@@ -0,0 +1,24 @@
+
+      (interaction regions [
+        (occlusion (-10,0) width=30 height=30)
+        (borderRadius 0.00),
```
this comma is .... weird?

hortont424 commented on Source/WebCore/page/DebugPageOverlays.cpp:
```diff
@@ -498,7 +498,7 @@ void InteractionRegionOverlay::drawRect(PageOverlay&, GraphicsContext& context,
     auto region = activeRegion();
 
-    if (region || !valueForSetting("constrain"_s)) {
+    if ((region && region->type != InteractionRegion::Type::Occlusion) || !valueForSetting("constrain"_s)) {
```
I think you want this check inside activeRegion() (i.e. never choose an occlusion region as active)

megangardner committed:
Small interaction regions need to be guarded with occlusion regions.
https://bugs.webkit.org/show_bug.cgi?id=254785
rdar://105092716

Reviewed by Tim Horton.

Also add some additional debug information to draw occlusion regions
in red.

* LayoutTests/interaction-region/input-type-file-region-expected.txt:
* LayoutTests/interaction-region/input-type-range-region-expected.txt:
* Source/WebCore/page/DebugPageOverlays.cpp:
(WebCore::InteractionRegionOverlay::drawRect):
* Source/WebCore/rendering/EventRegion.cpp:
(WebCore::EventRegionContext::guardSmallIntetactionRegions):
(WebCore::EventRegionContext::copyInteractionRegionsToEventRegion):
* Source/WebCore/rendering/EventRegion.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::paintDebugOverlays):

Canonical link: https://commits.webkit.org/262655@main
